### PR TITLE
feat: Add new ptn modules `avm/ptn/azd/ml-project`.

### LINKED ISSUE #1228: [Module Proposal]: `avm/ptn/azd/ml-project`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

Pattern Module

### Module Name

avm/ptn/azd/ml-project

### Module Details

Please add a Pattern Module named `avm/ptn/azd/ml-project`, this module will create the following three resources:

1. Machine Learning Services Workspace.
2. Key Vault.
3. User Assigned Identity.

For details, please refer to azd: 
https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/ai/project.bicep

CC: @jongio

### Do you want to be the owner of this module?

No

### Module Owner's GitHub Username (handle)

jongio

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by Menghua1**
## Description
Fixes https://github.com/Azure/Azure-Verified-Modules/issues/1228.

## Pipeline Reference

| Pipeline |
| -------- |
|   [avm.ptn.azd.ml-project](https://github.com/Menghua1/bicep-registry-modules/actions/workflows/avm.ptn.azd.ml-project.yml)     |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

@jongio for notification.

### CODE REVIEW

**Comment by Menghua1** [comment]
@AlexanderSehr Could you help review this PR?

**Comment by AlexanderSehr** [comment]
> @AlexanderSehr Could you help review this PR?

Hey @Menghua1,
as I'll be on vacation from this weekend for 2-ish weeks, it would probably make sense if another reviewer takes this and sees it through (cc: @Azure/avm-module-reviewers-bicep). Otherwise we may risk that I'll leave a comment and won't be able to resolve it before October 😄

**Comment by Menghua1** [comment]
> Hey @Menghua1, as I'll be on vacation from this weekend for 2-ish weeks, it would probably make sense if another reviewer takes this and sees it through (cc: @Azure/avm-module-reviewers-bicep). Otherwise we may risk that I'll leave a comment and won't be able to resolve it before October 😄

Hi AlexanderSehr,

Thank you for letting me know about your upcoming vacation. I understand the need to have another reviewer take over. Could you please help me ask if anyone from the @Azure/avm-module-reviewers-bicep group can review the PR promptly?

File: .github/workflows/avm.ptn.azd.ml-hub-dependencies.yml
Comment (eriqua) [inline .github/workflows/avm.ptn.azd.ml-hub-dependencies.yml:1]: Please remove as not part of this PR

File: avm/ptn/azd/ml-project/main.bicep [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      for id in roleDefinitionIdOrName: {
+        principalId: project.outputs.systemAssignedMIPrincipalId
+        roleDefinitionIdOrName: id
+        principalType: 'ServicePrincipal'
+      }
+    ]
+  }
+}
+
+resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.bicep:140]: Probably cosmetic, but imho it would help the readability of the module if resource and module declaration followed the deployment order

File: avm/ptn/azd/ml-project/main.bicep [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+param projectManagedIdentities managedIdentitiesType = {
+  systemAssigned: true
+}
+
+@description('Required. The name of the user assigned identity.')
+param userAssignedtName string
+
+@description('Optional. The role to assign. You can provide either the display name of the role definition, the role definition GUID, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
+param roleDefinitionIdOrName array = [
+  'f6c7c914-8db3-469d-8ca1-694a8f32e121'
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.bicep:71]: Since these are default values, I'd suggest mentioning the user friendly corresponding name in the description or at least as a comment close to the Id

File: avm/ptn/azd/ml-project/main.bicep [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+@description('Required. The name of the key vault.')
+param keyVaultName string
+
+@description('Optional. The managed identity definition for the machine learning resource. At least one identity type is required.')
+param projectManagedIdentities managedIdentitiesType = {
+  systemAssigned: true
+}
+
+@description('Required. The name of the user assigned identity.')
+param userAssignedtName string
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.bicep:67]: Please correct typo in all instances

File: avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+module keyvault 'br/public:avm/res/key-vault/vault:0.9.0' = {
+  name: '${uniqueString(deployment().name, location)}-keyvault'
+  params: {
+    enablePurgeProtection: false
+    location: location
+    name: keyVaultName
+  }
+}
+
+@description('The name of the AI Studio Hub Resource.')
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep:55]: Is this output used? If not I'd suggest to remove it for cleaner code

File: avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    tags: tags
+    location: location
+    sku: 'Basic'
+    kind: 'Hub'
+    associatedKeyVaultResourceId: keyvault.outputs.resourceId
+    associatedStorageAccountResourceId: storageAccount.outputs.resourceId
+  }
+}
+
+module storageAccount 'br/public:avm/res/storage/storage-account:0.9.1' = {
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep:37]: Also in this case it would help readability to order modules based on their order of deployment (e.g. put storage and kv before hub)

**Comment by ReneHezser** [comment]
@Menghua1 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

**Comment by Menghua1** [comment]
@eriqua I have updated the code based on your comments, please review it again, thank you!

**Comment by Menghua1** [comment]
> @Menghua1 We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

@ReneHezser Thank you for your guidance. I have updated branch and re-ran the Set-AVMModule utility.

File: .github/workflows/avm.ptn.azd.ml-ai-environment.yml
Comment (eriqua) [inline .github/workflows/avm.ptn.azd.ml-ai-environment.yml:1]: This workflow file should be removed since unrelated to this PR

File: avm/ptn/azd/ml-project/main.bicep [hunk 6]
```diff
@@ -0,0 +1,171 @@
+metadata name = 'Machine Learning workspace'
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.bicep:1]: Ca we please specify Azd in the readme title to avoid any confusion with the resource module?

File: avm/ptn/azd/ml-project/main.bicep [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+@description('Optional. The flag to signal HBI data in the workspace and reduce diagnostic data collected by the service.')
+param hbiWorkspace bool = false
+
+@description('Optional. Whether or not public network access is allowed for this machine learning resource. For security reasons it should be disabled.')
+@allowed([
+  'Enabled'
+  'Disabled'
+])
+param publicNetworkAccess string = 'Enabled'
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.bicep:53]: Is this intended to be enabled by default?

File: avm/ptn/azd/ml-project/main.json [hunk 8]
```diff
@@ -0,0 +1,7427 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "languageVersion": "2.0",
+  "contentVersion": "1.0.0.0",
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.28.1.47646",
```
Comment (eriqua) [inline avm/ptn/azd/ml-project/main.json:8]: Not mandatory, just a heads up that a newer version of bicep is available, in case you'd like to upgrade it

**Comment by eriqua** [comment]
> @eriqua I have updated the code based on your comments, please review it again, thank you!

Just a couple of additional comment @Menghua1, thank you for the great work

File: avm/ptn/azd/ml-project/main.bicep [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module project 'br/public:avm/res/machine-learning-services/workspace:0.7.0' = {
+  name: '${uniqueString(deployment().name, location)}-project'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-project/main.bicep:100]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-project/main.bicep [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
+  name: keyVaultName
+}
+
+module keyVaultAccess 'br/public:avm/res/key-vault/vault:0.9.0' = {
+  name: '${uniqueString(deployment().name, location)}-keyvault'
+  params: {
+    name: keyVault.name
+    accessPolicies: [
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-project/main.bicep:120]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-project/main.bicep [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        objectId: project.outputs.systemAssignedMIPrincipalId
+        permissions: { secrets: [ 'get', 'list' ] }
+      }
+    ]
+  }
+}
+
+module mlServiceRoleAssigned 'br/public:avm/res/managed-identity/user-assigned-identity:0.4.0' = {
+  name: '${uniqueString(deployment().name, location)}-roleassignment'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-project/main.bicep:131]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-project/main.bicep (hunk 7 above)
Comment (Menghua1) [reply avm/ptn/azd/ml-project/main.bicep:53]: Yes, it is aligned with the default values ​​of the [project.bicep](https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/ai/project.bicep#L17) file parameters in infra/core.

**Comment by Menghua1** [comment]
@eriqua and @AlexanderSehr  Thank you for your feedback! I have updated the code based on your comments, please review it again, thank you!

File: avm/ptn/azd/ml-project/main.bicep (hunk 9 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-project/main.bicep:100]: @Menghua1, please note, you did not add a passthru (`enableTelemetry: enableTelemetry`) but instead added the telemetry as a condition for the module's deployment (which you certainly don't want) 😉 

Please fix this for all module references. Also, I'd recommend to not resolve comments yourself, but simply comment them once you tackled them. Otherwise re-review are quite painful.

File: avm/ptn/azd/ml-project/main.bicep (hunk 10 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-project/main.bicep:120]: Not fixed (see other comment)

File: avm/ptn/azd/ml-project/main.bicep (hunk 11 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-project/main.bicep:131]: Not fixed (see other comment)

**Comment by Menghua1** [comment]
@AlexanderSehr I apologize for the oversight. Thank you for your feedback! I have submitted a new commit to address the issues correctly. Thank you very much.

**Comment by AlexanderSehr** [comment]
> @AlexanderSehr I apologize for the oversight. Thank you for your feedback! I have submitted a new commit to address the issues correctly. Please review it.

Much appreciated, thank you @Menghua1 . @eriqua will see the PR through.
One last note: Please resolve the conflicts 😉

**Comment by Menghua1** [comment]
> Much appreciated, thank you @Menghua1 . @eriqua will see the PR through. One last note: Please resolve the conflicts the merge of the `azd/ml-hub-dependencies` PR introduced to this PR 😉

OK, thank you for your guidance. @eriqua the conflict has been resolved. Please review it, thank you!😉

**Review by eriqua — APPROVED** [review]
🆗
